### DISCUSSION TITLE
Unexpected macOS curl CI breakage

### DISCUSSION THREAD

**Discussion by icing** [Homebrew/discussions#4643] [Everyday usage] [answered]
### Output of `brew config`

```shell
HOMEBREW_VERSION: >=2.5.0 (shallow or no git repository)
ORIGIN: (none)
HEAD: (none)
Last commit: never
Core tap: N/A
HOMEBREW_PREFIX: /usr/local
HOMEBREW_CASK_OPTS: []
HOMEBREW_DISPLAY: /private/tmp/com.apple.launchd.PrcjjZgqtH/org.macosforge.xquartz:0
HOMEBREW_MAKE_JOBS: 20
Homebrew Ruby: 2.6.10 => /usr/local/homebrew/Library/Homebrew/vendor/portable-ruby/2.6.10_1/bin/ruby
CPU: 20-core 64-bit cometlake
Clang: 14.0.3 build 1403
Git: 2.39.2 => /Library/Developer/CommandLineTools/usr/bin/git
Curl: 7.88.1 => /usr/bin/curl
macOS: 13.4.1-x86_64
CLT: 14.3.1.0.1.1683849156
Xcode: 14.3.1
```

### Output of `brew doctor`

```shell
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: You have unlinked kegs in your Cellar.
Leaving kegs unlinked can lead to build-trouble and cause formulae that depend on
those kegs to fail to run properly once built. Run `brew link` on these:
  openssl@3
```

### Description of issue

Some time in the last 2 weeks, CI jobs for curl on macOS started failing to build with TLS libraries other than the plain OpenSSL. Today, after a brew upgrade, I reproduced the problem on my on machine.

We would like to understand the problem and what did change. It would be nice if someone here could help with that. We have a workaround by `brew unlink openssl` now.

More details:
- on the CI, we use brew to install libraries needed by curl. This varies per CI job. Some install openssl and libssh2, some install libressl and libssh2, etc. This all worked fine for some time now.
- the CI jobs with libressl started failing. Analysis showed that the curl build used the include files from `/usr/local/include/openssl` instead of the libressl ones. It then tried to link against libressl and symbols were missing.
- The additional include dirs in curl build are added using `-isystem` to avoid warnings from those headers.
- In effect, the build called `gcc -isystem /usr/local/libressl/include -v source.c`, but it got the openssl header. That was surprising.
- Looking at `echo | gcc -E -Wp,-v -` shows:

```
clang -cc1 version 14.0.3 (clang-1403.0.22.14.1) default target x86_64-apple-darwin22.5.0
ignoring nonexistent directory "/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/local/include"
ignoring nonexistent directory "/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/Library/Frameworks"
#include "..." search starts here:
#include <...> search starts here:
 /usr/local/include
 /Library/Developer/CommandLineTools/usr/lib/clang/14.0.3/include
 /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include
 /Library/Developer/CommandLineTools/usr/include
 /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/System/Library/Frameworks (framework directory)
End of search list.
```

So `/usr/local/include` is at the start of the system paths, but anything added with `-isystem` should be looked at before that. But it clearly is not. This is very strange.

By `brew unlink openssl` we avoid now this borked include path order, but we sure would like to know what changed. Thanks for any help you might provide.

**Comment by carlocab** [comment]
Could you link to an example failing CI job? Is the error the same across failures?

**Reply by icing** [reply]
Sure, for example [this one](https://github.com/curl/curl/actions/runs/5499059260/jobs/10020987667). The linker error is always the same, because the openssl includes trigger`#idef'ed` use of functions not available in libressl.

**Reply by carlocab** [reply, marked as answer]
I think you're experiencing this because we recently decided to link `openssl@3` into `HOMEBREW_PREFIX` at https://github.com/Homebrew/homebrew-core/commit/501a8a714dbb6663158344acd43c3c1c1511ff95. See https://github.com/Homebrew/homebrew-core/issues/104784 for discussion and reasoning.

Essentially this means that OpenSSL 3.1 headers can now be found at `/usr/local/include/openssl` whenever it's installed.

As you've found, doing `brew unlink openssl` is a suitable solution, because this removes `openssl@3` from `/usr/local`.

Here are some alternative workarounds:

1. After [this step](https://github.com/curl/curl/blob/9630fb9f2a4056805e787f0a71e2d86991f9cd20/.github/workflows/macos.yml#L116-L120) in your CI workflow, do `brew bundle cleanup --force --file=/tmp/Brewfile`. What this does is uninstalls all formulae that are not listed in your `Brewfile` so they don't interfere with your build/tests.
2. Run `brew test-bot --only-cleanup-before` _before_ doing `brew bundle install`. This also does something equivalent to the first suggestion above, but also does some extra cleanup of the GitHub runner (see details [here](https://github.com/Homebrew/homebrew-test-bot/blob/0850fba8366874854b6aff7d1c8393bd78ddb108/lib/tests/cleanup_before.rb)). These are cleanup steps that we do before every build at Homebrew to make sure that software that is pre-installed on the GitHub runner don't interfere with our builds.

If you choose to do the second suggestion above, you'll need to either run `brew tap homebrew/test-bot` or use [Homebrew/actions/setup-homebrew](https://github.com/Homebrew/actions/tree/master/setup-homebrew) with something like
```yaml
- uses: Homebrew/actions/setup-homebrew@dccbb994e6e28406006e81ab3ba5ea1076a36848
  with:
    test-bot: true
```
to make sure that the `brew test-bot` command is available.

Or you can continue to do `brew unlink openssl` as you're already doing.

Apologies for the breakage! Feel free to tag me if you encounter any other Homebrew-related issues, or otherwise file an issue at Homebrew/homebrew-core (for Homebrew formulae/package bugs) or Homebrew/brew (for `brew` bugs).

**Comment by icing** [comment]
Thanks for the quick answer and the proposed actions for fixing. I'll think I make an attempt with approach 1. So far, we had no interference with things github has already added, but it is good to know that there is a solution to reset that, should we need it.

**Comment by bagder** [comment]
The main problem will be for the poor people who want to install and build with OpenSSL as well as one of the OpenSSL forks, since this change makes the OpenSSL includes overly sticky and really hard to avoid. It seems like a bug to me.

**Reply by Bo98** [reply]
In theory it should work as in many cases as with Linux where you often have a system OpenSSL installed and but you can still have your own build when you pass the correct include paths to that. Homebrew isn't doing anything special here, with the only quirk being that some build scripts add in the Homebrew include path because it's not included by the compiler by default. Anyone doing this should however do it after any other include flags (or use `-isystem`).

I wonder if there's something that's elevating the Homebrew include path or adding it really early? If someone can minimise the steps to reproduce it might make the issue more obvious.

`brew unlink openssl` or any of the other workarounds should never be necessary IMO, so we should track down what is adding the include path. It's not really OpenSSL specific either given I imagine it could equally happen if you have a custom version of any other library.

**Comment by Bo98** [comment]
Ok looked into this properly and I see what's happening now I think:

> * In effect, the build called `gcc -isystem /usr/local/libressl/include -v source.c`, but it got the openssl header. That was surprising.
> * Looking at `echo | gcc -E -Wp,-v -` shows:
> ```
> clang -cc1 version 14.0.3 (clang-1403.0.22.14.1) default target x86_64-apple-darwin22.5.0
> ignoring nonexistent directory "/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/local/include"
> ignoring nonexistent directory "/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/Library/Frameworks"
> #include "..." search starts here:
> #include <...> search starts here:
>  /usr/local/include
>  /Library/Developer/CommandLineTools/usr/lib/clang/14.0.3/include
>  /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include
>  /Library/Developer/CommandLineTools/usr/include
>  /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/System/Library/Frameworks (framework directory)
> End of search list.
> ```

This seems to be a quirk specific to `/usr/local` install (so not arm64 Homebrew) and how Apple's clang works when run via `/usr/bin/clang` and `/usr/bin/gcc`. These basically run through `xcrun` and `xcrun` does some awful trickery:

```
$ xcrun sh -c 'echo $CPATH'
/usr/local/include
```

I'm not sure why it does this, as invoking the compiler directly will still have `/usr/local/include` but in its correct position as a system include.

`-I` should however be unaffected and still take priority, and I generally recommend using that instead of `-isystem` unless you're actually dealing with system headers you want to reduce the priority of, which does not seem to be the case here based on your issue. Both apply to `""` and `<>` includes. `-I` should solve your issue I believe - if not let me know and I can dig deeper. Most pkg-config files also use `-I`.
